rust/src/bs_safe.rs

```diff
@@ -0,0 +1,535 @@
+        let n = self.param_unc_num();
+        assert_eq!(
+            theta_unc.len(),
+            n,
+            "Argument 'theta_unc' must be the same size as the number of parameters!"
+        );
+        assert_eq!(
+            grad.len(),
+            n,
+            "Argument 'grad' must be the same size as the number of parameters!"
+        );
+
+        let mut val = 0.0;
+
+        let mut err = ErrorMsg::new(self.lib.borrow());
+        let rc = unsafe {
+            self.lib.borrow().0.bs_log_density_gradient(
+                self.model.as_ptr(),
+                propto as c_int,
+                jacobian as c_int,
+                theta_unc.as_ptr(),
+                &mut val,
+                grad.as_mut_ptr(),
+                err.as_ptr(),
+            )
+        };
+
+        if rc == 0 {
+            Ok(val)
+        } else {
+            Err(BridgeStanError::EvaluationFailed(err.message()))
+        }
+    }
+
+    /// Compute the log of the prior times likelihood density and gradient and hessian.
+    ///
+    /// Drop jacobian determinant terms if `jacobian == false` and
+    /// drop constant terms of the density if `propto == true`.
+    /// The gradient of the log density is stored in `grad`, the
+    /// hessian is stored in `hessian`.
+    pub fn log_density_hessian(
+        &self,
+        theta_unc: &[f64],
+        propto: bool,
+        jacobian: bool,
+        grad: &mut [f64],
+        hessian: &mut [f64],
+    ) -> Result<f64> {
+        let n = self.param_unc_num();
+        assert_eq!(
+            theta_unc.len(),
+            n,
+            "Argument 'theta_unc' must be the same size as the number of parameters!"
+        );
+        assert_eq!(
+            grad.len(),
+            n,
+            "Argument 'grad' must be the same size as the number of parameters!"
+        );
+        assert_eq!(
+            hessian.len(),
+            n.checked_mul(n).expect("Overflow for size of hessian"),
+            "Argument 'hessian' must be the same size as the number of parameters squared!"
+        );
+
+        let mut val = 0.0;
+
+        let mut err = ErrorMsg::new(self.lib.borrow());
+        let rc = unsafe {
+            self.lib.borrow().0.bs_log_density_hessian(
+                self.model.as_ptr(),
+                propto as c_int,
+                jacobian as c_int,
+                theta_unc.as_ptr(),
+                &mut val,
+                grad.as_mut_ptr(),
+                hessian.as_mut_ptr(),
+                err.as_ptr(),
+            )
+        };
+
+        if rc == 0 {
+            Ok(val)
+        } else {
+            Err(BridgeStanError::EvaluationFailed(err.message()))
+        }
+    }
+
+    /// Map a point in unconstrained parameter space to the constrained space
+    pub fn param_constrain<R: Borrow<StanLibrary>>(
+        &self,
+        theta_unc: &[f64],
+        include_tp: bool,
+        include_gq: bool,
+        out: &mut [f64],
+        rng: &mut Rng<R>,
```

Comment:
If we wanted to match the other languages I think this would be an `Option` type, which we require to be `Some` when `include_gq` is true.

I'd also accept that in a lower-level binding we'd prefer the safety of just always having it rather than doing that sort of dynamic check, but I just wanted to raise it for discussion.

Reply:
I changed it to an Option<rng> and added a runtime check that it has to be `Some` if `include_gq`.